Attachment Handling

thororm opened the pull request:
<!-- This is a template for your Pull Request. This are just some suggestions for you. You do not have to use all of them. -->

<!-- If your PR fixes an issue, mention it here. You can also just copy the URL - GitHub will convert it for you.
If this PR fixes several issues, please prepend each issue url/number with the word "fix"/"fixes" or "close"/"closes" as this automatically closes the issues you mentioned when the PR is merged.
-->
This PR fixes https://github.com/PrivateBin/PrivateBin/issues/14

## Changes
<!-- List all the changes you have done -->
Added display of
- Video (HTML5)
- Audio (HTML5)
- PDF (Browser capabilities are used for this)

attachments.

Added drag & drop functionality
Added attachment display to preview before submitting

I had to change the cspheader for these functionalities to work.

rugk commented:
Thanks for this, this is a very good idea. :+1: 

I am just currently refactoring the JS code in https://github.com/PrivateBin/PrivateBin/pull/180, so you will need to adjust/integrate it into the new code when the refactoring is merged.
So you may have a look at that PR to see what changed and how you can integrate your PR there.

rugk commented on cfg/conf.ini.sample:
```diff
@@ -64,7 +64,7 @@ languageselection = false
 ; Check the documentation at https://content-security-policy.com/
 ; Note: If you use a bootstrap theme, you can remove the allow-popups from the sandbox restrictions.
-; cspheader = "default-src 'none'; manifest-src 'self'; connect-src *; script-src 'self'; style-src 'self'; font-src 'self'; img-src 'self' data:; referrer no-referrer; sandbox allow-same-origin allow-scripts allow-forms allow-popups"
+; cspheader = "default-src 'none'; manifest-src 'self'; connect-src *; script-src 'self'; style-src 'self'; font-src 'self'; img-src 'self' data:; media-src 'self' data:; object-src  'self' data:; Referrer-Policy: 'no-referrer'"
```
You don't seem to need `'self'` as you are only doing it locally via `data:`.

rugk commented:
Also as for PDF I've heard that Chrome/ium might apply the CSP to the internal PDF viewer too, so this is certainly a thing we need to test.

rugk commented on js/privatebin.js:
```diff
@@ -340,6 +340,180 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+            }
+
+            function drop(e) {
+                e.stopPropagation();
```
Please just use a readable variable name like `event`. If someone wants to minify the JS they can do so, but the code here should be readable. I also think some code checkers would complain…

thororm commented:
Hi rugk,

I updated the PR.
It also works without self in the csp.

Yes, Chrome does apply the csp header to the pdf viewer. It treats it like any other plugin and doesn't allow it in a sandboxed environment. Unfortunately, there is currently no way to allow the PDF viewer in this case: https://bugs.chromium.org/p/chromium/issues/detail?id=413851 (An iFrame also did not work.)
In Firefox the display of PDFs also works with the sandbox header.
Maybe pdf.js could be an alternative, but I don't think it's that easy to integrate with a good PDF UI. And due to it's size, it should only be loaded if necessary.

I will take a look at the refactored version and try to integrate the functionality there as soon as it's finished.

rugk commented on lib/Configuration.php:
```diff
@@ -51,7 +51,7 @@ class Configuration
             'urlshortener'             => '',
             'icon'                     => 'identicon',
-            'cspheader'                => 'default-src \'none\'; manifest-src \'self\'; connect-src *; script-src \'self\'; style-src \'self\'; font-src \'self\'; img-src \'self\' data:; media-src \'self\' data:; object-src \'self\' data:; Referrer-Policy: \'no-referrer\'',
+            'cspheader'                => 'default-src \'none\'; manifest-src \'self\'; connect-src *; script-src \'self\'; style-src \'self\'; font-src \'self\'; img-src \'self\' data:; media-src data:; object-src data:; Referrer-Policy: \'no-referrer\'',
```
Also do you really think object-src is necessary. This usually just included flash and other plugins… Do PDF readers belong to this? Is there a difference between Chrome & Firefox regarding this?

rugk replied:
And BTW, also adjust the CSP in the config sample.

thororm replied:
Yes, object-src is required by Chrome and Firefox for the PDF view to work

thororm replied:
Thanks, I forgot to update the sample

rugk commented:
> Maybe pdf.js could be an alternative

I also think this would be bad.
And if Chrome's PDF viewer would work without the sandbox in the CSP this would be nice (as we could remove it - the sandbox is not very useful anyway currently), but I think some other directives would potentially break it…

thororm commented:
Chromes PDF viewer does work without the sandbox header. Firefox works with and without the header.
The current version works in Chrome and Firefox.

rugk commented:
So just remove the sandbox value then.

thororm commented:
It is already removed

rugk commented:
I've basically completed refactoring. There will no more big changes in https://github.com/PrivateBin/PrivateBin/issues/93%E2%80%A6, so you can have a look there…

rugk commented:
@thororm Are you still interested in completing this PR?

thororm commented:
@rugk: Yes, I just didn't have the time to take a look at the refactored version yet. Hopefully I can do that soon

elrido commented:
It took longer then expected, but the JS refactoring is now finally complete.

rugk commented:
@thororm I am willing to take this over at the end of next week or beginning of the week afterwards. If you don't mind I'll use your source code and include it in the current master. Of course, you can also continue working on it, but going forward we would like to integrate this feature here, soon.

thororm commented:
@elrido, @rugk: I finally found some time to integrate the attachment handling into the refactored version.

rugk commented:
Please don't put this attachmentHelpers object under the Helpers object. (we don't want to have such nested things) These functions may also not really be helpers, but actually important functions for displaying an attachment so put them in the `AttachmentViewer` object.

thororm commented:
You're right, I moved the functions to the AttachmentViewer.

How should we handle the failing testMessageIdsExistInAllLanguages test?
Add the english translation as a dummy?

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+                // revert loading status…
+                AttachmentViewer.hideAttachment();
+                AttachmentViewer.hideAttachmentPreview();
+                Alert.showError(I18n._('Your browser does not support uploading encrypted files. Please use a newer browser.'));
```
Please remove the `I18n.`. The new version of this function automatically translates the text.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+            var fileReader = new FileReader();
+            if (file === undefined) {
+                file = me.$fileInput[0].files[0];
+                $('#dragAndDropFileName').text('');
```
Maybe preload this in a `me.$dragDropFileName` variable or so?

rugk commented on js/privatebin.js:
```diff
@@ -2257,6 +2088,9 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
          */
         me.hasAttachment = function()
         {
+            if (!AttachmentViewer.$attachment.length) {
```
If you are in `AttachmentViewer` you can just use `me` instead of referring to it with its full name.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+                me.attachmentData = dataURL;
+
+                if (Editor.isPreview()) {
+                    me.handleAttachmentPreview(AttachmentViewer.$attachmentPreview, dataURL);
```
If you are in `AttachmentViewer` you can just use `me` instead of referring to it with its full name.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+                    this.showPDFPreview(targetElement, data);
+                }
+                //else {
+                //console.log("file but no image/video/audio/pdf");
```
I think you can remove that commented code.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+         */
+        me.handleAttachmentPreview = function (targetElement, data) {
+            if (data) {
+                var mimeType = this.getMimeTypeFromDataURL(data);
```
Instead of this, use `me`. Also below.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+         * Get Mime Type from a DataURL
+         *
+         * @param {type} dataURL
+         * @returns Mime Type from a dataURL as obtained for a file using the FileReader API https://developer.mozilla.org/en-US/docs/Web/API/FileReader#readAsDataURL()
```
Maybe rather use the link in a `@see` line.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+            var videoPlayer = $(document.createElement('video'))
+                    .attr('controls', 'true')
+                    .attr('autoplay', 'true')
+                    .attr('loop', 'true')
```
Err, I think we usually do not want to have looping videos by default.

rugk replied:
And maybe also autoplay is not that nice, but this can be discussed.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+            targetElement.removeClass('hidden');
+        };
+
+        me.showVideoPreview = function (targetElement, video, mimeType) {
```
As targetElement is a jQuery element prepend the variable with `$`. So: `$targetElement`.
Also JSDOC would be nice.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+        };
+
+        me.showVideoPreview = function (targetElement, video, mimeType) {
+            var videoPlayer = $(document.createElement('video'))
```
Same here. This seems to be a jQuery element.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+
+            //Firefox crashes with files that are about 1.5MB
+            //The performance with 1MB files is bareable
+            if (pdf.length < 1398488) {
```
Maybe make an early return (length > … --> return) instead of this big indentation. It just looks better…

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+                        );
+                targetElement.removeClass('hidden');
+            } else {
+                Alert.showError(I18n._('File too large, to display a preview. Please download the attachment.'));
```
Again without I18n.

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+         * @returns {undefined}
+         */
+        me.addDragDropHandler = function () {
+            var fileInput = me.$fileInput;
```
Why save it in a local var? Just use the var as it…

rugk commented on js/privatebin.js:
```diff
@@ -2295,6 +2129,192 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+         * @returns {undefined}
+         */
+        me.addClipboardEventHandler = function () {
+            document.addEventListener("paste",
```
Would not jQuery's `on()` method also work?

rugk commented on js/privatebin.js:
```diff
@@ -3917,10 +3943,12 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
-                    // missing) return to let JS continue and wait for user
-                    if (!decryptAttachment(paste, key, password)) {
-                        return;
+                    if (AttachmentViewer.$attachment.length) {
```
If you are in AttachmentViewer, you can use `me`. Otherwise this should not work as variables should not be exposed publicly in modules. Create a getter method if needed.

rugk commented:
> Add the english translation as a dummy?

Yes, in all languages, of course. When this PR is merged we should comment in https://github.com/PrivateBin/PrivateBin/issues/201.

thororm commented:
Thanks for the review. I updated the code.

I removed the loop from the video, but would like to keep the autoplay, as it is also the default for most video sites. (youtube etc...)

rugk commented on js/privatebin.js:
```diff
@@ -2123,149 +2136,162 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+                $($dragAndDropFileName).text('');
             } else {
-                $('#dragAndDropFileName').text(file.name);
+                $($dragAndDropFileName).text(aFile.name);
```
`$dragAndDropFileName` should either be already be a jQuery object so it should already be set (in an `init()` method) – I prefer this – or you use a method without prepending the var with `$`, which is actually a string: `'#dragAndDropFileName'`.

So $ is a determinator for a jQuery object, so you must not wrap it in `$(…)` in another time.

rugk commented on js/privatebin.js:
```diff
@@ -2123,149 +2136,162 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
         /**
          *  Handle the preview of files that can either be an image, video, audio or pdf element.
-         *  @argument {DOM Element} targetElement where the preview should be appended.
+         *  @argument {DOM Element} $targetElement where the preview should be appended.
```
Upps, sorry. If the element is really only a DOM element and not a jQuery element, you must of course not prepend it with $. Maybe, I've misunderstood the code here, sorry.
Currently the JSDOC here is a bit contrary: `DOM Element <-> $…`

If I see it correctly better document a jQuery object – if it is one! – as `{jQuery}`.

rugk commented on js/privatebin.js:
```diff
@@ -2123,149 +2136,162 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
          * @param {type} dataURL
-         * @returns Mime Type from a dataURL as obtained for a file using the FileReader API https://developer.mozilla.org/en-US/docs/Web/API/FileReader#readAsDataURL()
+         * @see {@link https://developer.mozilla.org/en-US/docs/Web/API/FileReader#readAsDataURL()}
+         * @returns Mime Type from a dataURL as obtained for a file using the FileReader API
```
First, define a data type there. (string, yeah?)
Ref: http://usejsdoc.org/tags-returns.html

rugk commented on js/privatebin.js:
```diff
@@ -2123,149 +2136,162 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
-            targetElement.html(
+        /**
+         * Displays an image attachment within the $targetElement.
+         * @param {type} $targetElement jQuery reference to the element, where the image should be displayed.
```
Specify `{type}`, please. Here `jQuery` e.g.

rugk commented on js/privatebin.js:
```diff
@@ -2123,149 +2136,162 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
+                return;
+            }
 
                 //Fallback for browsers, that don't support the vh unit
```
De-intend, or?
You seems to have missed removing the indentation.

rugk commented:
Thank for the changes, I've approved the code now. Still needs some testing and a code review by @elrido, but afterwards we can merge this.
